include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -3303,6 +3303,31 @@ struct OrtApi {
   */
   ORT_API2_STATUS(SessionOptionsAppendExecutionProvider_MIGraphX,
                   _In_ OrtSessionOptions* options, _In_ const OrtMIGraphXProviderOptions* migraphx_options);
+
+ /** \brief Replace initialized Tensors with external data with the data provided in initializers.
+  *
+  * The function will find the initialized TensorProtos with external data in the graph with the provided names and
+  * replace them with the provided tensors. The API verifies that the TensorProto being replaced
+  * has an external data reference and has the same name, dimensions and data type as its replacement. The replacement
+  * will occur before any of the optimizations take place. The data will be copied into the graph
+  * since TensorProto can't refer to the user provided buffers.
+  *
```

Comment:
(sorry, didn't notice that the previous thread above closed - reposting comment in case it isn't visible) 

Note that the proposal is not to change the tensor protos that reside within the onnx file itself but the in-memory structures created by ORT to track them after the model is loaded (structures that can then be made available to downstream code).

Are modifiable initializers a recent change to optimization? Under release 1.7 and earlier, external data initializers were mmap'ed and the data was used directly, uncopied and apparently without manipulation. Subsequent releases disabled that as noted in issue 10977 which PR 11127 tries to address (the tested citrinet model referenced in the PR is read-only). However, optimizers that can manipulate (external) initializers could throw a wrench in mmap use cases that aim to reduce load time and/or peak memory usage.

Reply:
I have seen your comment and I have seen your PR. I can't answer the question as to when the specific optimizers were introduced, however, they can manipulate graph structures at will and they should gain the best performance at inference time. At the same time, we support saving an optimized graph. What you are referring to is a subject of a design investigation.

Reply:
While working on PR 11127 I did notice that transformations/optimizations can be applied during the load. However, these appear limited to the graph structure - as opposed to the content of the initializers - since they occur before the call to `SaveInitializedTensors` which actually loads the external data.

This seems consistent with what I observe when using `SessionOptions::SetOptimizedModelFilePath`... if I enable optimizations a new `onnx` file is created but the external data files are untouched (and no new ones are created).